### PR TITLE
All Antenna Sky-Subtraction 2 Methods

### PR DESCRIPTION
**Pull request by tamerakassie**
Added the notebook that compares the ABBA and the  2 channel Time Differencing Sky Subtraction Methods, also updated with the bandpass plots

### CODE REVIEW

**Comment by tamerakassie** [comment]
<p>okay, let me know how we can do this</p><p> </p>

---

 <a href='https://app.reviewnb.com/meerklass/meerklass-ssins/pull/6/discussion/#comment-1980468318'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by tamerakassie** [comment]
<p>just a question, will it be a good idea to plot all  receiver waterfalls ? or just for a few? </p>

---

 <a href='https://app.reviewnb.com/meerklass/meerklass-ssins/pull/6/discussion/#comment-1980468559'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by piyanatk** [comment]
<p>All receivers would be nice, but it might be too many plots. You will probably have to combine them into one big multi-panel plots, say 13 rows x 5 columns. Something like</p><pre class="ql-syntax" spellcheck="false"># Normalisation for all panels
norm = LogNorm(min= ..., max= ...)

ncols = 5
fig, axes = plt.subplots(13, ncols, figsize=(15, 40), sharex='all', sharey='all)
# Loop over antennas, plotting with <code>plot_waterfall</code>
for i, ant in enumerate(axes.flat):
    # Check if i is in good antenna list. If so, plot. Otherwise, left the subplot empty
    if i in good_ants_int:
        plot_waterfall(data, ax=ax)
fig.colorbar(axes)
</pre>

---

 <a href='https://app.reviewnb.com/meerklass/meerklass-ssins/pull/6/discussion/#comment-1980468559'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by piyanatk** [comment]
<p><a href="https://stackoverflow.com/questions/14463277/how-to-disable-python-warnings/14463362#14463362" target="_blank">https://stackoverflow.com/questions/14463277/how-to-disable-python-warnings/14463362#14463362</a></p>

---

 <a href='https://app.reviewnb.com/meerklass/meerklass-ssins/pull/6/discussion/#comment-1980468318'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: mkssins.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+#Imports
+import numpy as np
+import matplotlib.pylab as plt
+import warnings
+import time
+import pickle
+import sys
+import katcali
```
Comment (piyanatk) [inline mkssins.py]: It doesn't look any of these are needed in the module file (here). Remove them and import in the notebook instead if you need them there.

File: mkssins.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    for i in np.array(kio.ant_list(data)):
+        if i not in bad_ants:
+            ants_good.append(i)
+    else:
+        print (str(i) + ' is bad')
+
+    return ants_good
+"""
+Define a function that will access the data block and return visibility data, flags, noise diodes vector. 
+"""
```
Comment (piyanatk) [inline mkssins.py]: Docstring should go after the `def` line.

File: mkssins.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+import sys
+import katcali
+import katcali.io as kio
+import katcali.label_dump as kl
+import katcali.diode as kd
+from pathlib import Path
+def good_ant(fname):
+
+    """
+    Fuction to retrieve list of good antennas.
```
Comment (piyanatk) [inline mkssins.py]: Add `fname` to docstring

File: mkssins.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    return ants_good
+"""
+Define a function that will access the data block and return visibility data, flags, noise diodes vector. 
+"""
+def visData(fname, ant, pol):
+    """
+    Parameters:
+    ----------
+    Fname: Path to observation block.
```
Comment (piyanatk) [inline mkssins.py:44]: add parameters to docstring

File: mkssins.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+    Note : Current function looks at one pol and one dish.
+    """
+
+    data = kio.load_data(fname)
+    target, c0, band_ants, flux_model = kio.check_ants(fname)
+    ants_good = good_ant(fname)
+    data.select(ants=ant, pol=pol)
+    recv = ant + pol
+    print(recv)
```
Comment (piyanatk) [inline mkssins.py]: You probably won't need this printed out. I suggest adding and option `verbos=False` to the function definition. Then, here you can do
```
if verbos:  # i.e. verbos=True
    print(recv)
```
So you can make it print information for debugging when needed.

File: mkssins.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    corr_id = kio.cal_corr_id(data, recv)
+
+    assert(recv == data.corr_products[corr_id][0])
+    assert(recv == data.corr_products[corr_id][1])
+
+    print("Correlation ID:", corr_id, "Receiver:", recv)
```
Comment (piyanatk) [inline mkssins.py]: Same as above

File: mkssins.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    # Load visibilities and flags
+    vis, flags = kio.call_vis(fname, recv)
+    print("Shape of vis:", vis.copy().shape)
```
Comment (piyanatk) [inline mkssins.py]: Save as above

File: mkssins.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    nd_on_edge, nd_off_edge = kd.cal_nd_edges(timestamps, nd_set, nd_cycle, nd_on_time)
+    nd_ratio, nd_0, nd_1x = kd.cal_nd_ratio(timestamps, nd_on_time, nd_on_edge, data.dump_period)
+
+
+    nd_t0, nd_t1x, nd_s0, nd_s1x, nd_t0_ca, nd_t0_cb, nd_t1x_ca, nd_t1x_cb = kl.cal_label_intersec(dp_tt, dp_ss, nd_0, nd_1x)
+
+    return vis, nd_s0
+
+
+def MaskedArrayVisibilityFlags(vis, flags_mask, nd_s0):
```
Comment (piyanatk) [inline mkssins.py]: Probably call `flags_mask` just `flags` for clarity

File: mkssins.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    out: ndarray after ABBA (shape N-3)
+    """
+    array = array.filled()
+    vis_abba = (array[1:-2]+array[2:-1])/2 - (array[0:-3]+array[3:])/2
+    return vis_abba
+
+
+
+
+def plot_bandpass(x : np.ndarray, y : np.ndarray, label=None, Title =None, ylim : tuple = None, figsize=(8, 3), ax=None):
```
Comment (piyanatk) [inline mkssins.py]: Probably combine this with `plot`

**Review by piyanatk — APPROVED** [review]
@tamerakassie This looks good for now. I have more suggestions on how to tidy this up more, but we can chat about that on Wednesday.

**Pull request merged**